### PR TITLE
AMDGPU: Refactor atomicrmw fadd expansion logic (NFC)

### PR DESCRIPTION
**Pull request by arsenm**
This had some repeated and overlapping conditions, which
made it more difficult to handle the new metadata scheme. Reflow
the function to handle the easy LDS cases first. For the flat/global
cases, write in a positive-enabled style where everything unhandled
hits a default cmpxchg.

Depends #89468

### CODE REVIEW

**Review by Pierre-vh — APPROVED** [review]
Is this NFC(I)? If so please add it to the commit title

File: llvm/lib/Target/AMDGPU/SIISelLowering.cpp [hunk 1]
```diff
@@ -16075,56 +16075,50 @@ SITargetLowering::shouldExpandAtomicRMWInIR(AtomicRMWInst *RMW) const {
...
+        AS != AMDGPUAS::BUFFER_FAT_POINTER)
       return AtomicExpansionKind::CmpXChg;
 
-    if ((AMDGPU::isFlatGlobalAddrSpace(AS) ||
-         AS == AMDGPUAS::BUFFER_FAT_POINTER) &&
-        Subtarget->hasAtomicFaddNoRtnInsts()) {
-      if (Subtarget->hasGFX940Insts())
-        return AtomicExpansionKind::None;
+    // TODO: gfx940 supports v2f16 and v2bf16
+    if (Subtarget->hasGFX940Insts() && (Ty->isFloatTy() || Ty->isDoubleTy()))
```
Comment (rampitec) [inline llvm/lib/Target/AMDGPU/SIISelLowering.cpp:16083]: I lost the track of this condition after many changes. Can this hasGFX940Insts() be spelled out in terms of a specific feature/target property or its combination? What about gfx11/12, we seem to expand even f32?

File: llvm/lib/Target/AMDGPU/SIISelLowering.cpp [hunk 2]
```diff
@@ -16075,56 +16075,50 @@ SITargetLowering::shouldExpandAtomicRMWInIR(AtomicRMWInst *RMW) const {
...
+      if (RMW->use_empty() && Subtarget->hasAtomicFaddNoRtnInsts())
+        return ReportUnsafeHWInst(AtomicExpansionKind::None);
+      // global/buffer atomic fadd f32 rtn: gfx90a, gfx940, gfx11+.
+      if (!RMW->use_empty() && Subtarget->hasAtomicFaddRtnInsts())
         return ReportUnsafeHWInst(AtomicExpansionKind::None);
+    }
 
-      // global and flat atomic fadd f64: gfx90a, gfx940.
-      if (Ty->isDoubleTy() && Subtarget->hasGFX90AInsts())
+    // flat atomic fadd f32: gfx940, gfx11+.
```
Comment (rampitec) [inline llvm/lib/Target/AMDGPU/SIISelLowering.cpp:16106]: Comment seems to be wrong, gfx940 shall be already handled by this point (where the check for hasGFX940Insts is).

File: llvm/lib/Target/AMDGPU/SIISelLowering.cpp (hunk 1 above)
Comment (rampitec) [reply llvm/lib/Target/AMDGPU/SIISelLowering.cpp:16083]: Should that be something like:
```
if  (Subtarget->hasAgentScopeFineGrainedRemoteMemoryAtomics() &&
     Subtarget->hasMemoryAtomicFaddF32DenormalSupport()) {
  if (Subtarget->hasAtomicFaddRtnInsts() && Ty->isFloatTy())
    return AtomicExpansionKind::None;
  if (Subtarget->hasFlatBufferGlobalAtomicFaddF64Inst() && Ty->isDoubleTy())
    return AtomicExpansionKind::None;
}
```
Maybe also check addresspace global vs flat.

I believe that was the original intent of the gfx940 check, but I am not sure the meaning of unsafe is still the same.

Comment (rampitec) [reply llvm/lib/Target/AMDGPU/SIISelLowering.cpp:16083]: I have created https://github.com/llvm/llvm-project/pull/100217

**Pull request merged**